pylint/checkers/refactoring/recommendation_checker.py

```diff
@@ -1,12 +1,40 @@
 # Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 # For details: https://github.com/PyCQA/pylint/blob/master/LICENSE
 
+from typing import Optional, Union, cast
+
 import astroid
 
 from pylint import checkers, interfaces
 from pylint.checkers import utils
 
 
+def _check_if_dict_keys_used(
+    node: Union[astroid.For, astroid.Comprehension]
+) -> Optional[str]:
+    if not isinstance(node.iter, astroid.Call):
+        # Is it a dictionary?
+        if not isinstance(node.iter, (astroid.Name, astroid.Attribute)):
+            return None
+        inferred = utils.safe_infer(node.iter)
+        if not isinstance(inferred, (astroid.Dict, astroid.DictComp)):
+            return None
+        iterating_object_name = node.iter.as_string()
+
+    else:
+        # Is it a proper keys call?
+        if (
+            isinstance(node.iter.func, astroid.Name)
+            or node.iter.func.attrname != "keys"
+        ):
+            return None
+        inferred = utils.safe_infer(node.iter.func)
+        if not isinstance(inferred, (astroid.BoundMethod, astroid.Dict)):
+            return None
+        iterating_object_name = node.iter.as_string().partition(".")[0]
+    return iterating_object_name
```

Comment:
Since this is it's own function now, we can probably refactor it.
- Replace `iterating_object_name = ...` calls with `return ...`
- I would prefer _positive_ `if` statements, so switch `if not` and `else` to `if isinstance(node.iter, astroid.Call)` and `else`. (The `else` then is probably unnecessary, since the `if` block is closed via a `return`.)